### ISSUE TITLE
Score data being lost on pickling and unpickling

### ISSUE THREAD

**Issue by appukuttan-shailesh** [scidash/sciunit#206]
@rgerkin any thoughts on why I seem to lose the `related_data` info on pickling and unpickling:

```python
from sciunit.scores.complete import ZScore
score = ZScore(1.2)
print(score)
print(type(score))
print(score.related_data)
score.related_data["figures"] = ["a"]
print(score.related_data)
print(score.__dict__.keys())
```

>Z = 1.20
<class 'sciunit.scores.complete.ZScore'>
{}
{'figures': ['a']}
dict_keys(['score', 'related_data'])

```python
import pickle
with open('sample.pkl', 'wb') as f1:
    pickle.dump(score, f1)

with open('sample.pkl', 'rb') as f2:
    data = pickle.load(f2)

print(data)
print(type(data))
print(data.related_data)
print(data.__dict__.keys())
```

>Z = 1.20
<class 'sciunit.scores.complete.ZScore'>
None
dict_keys(['description', 'log10_norm_score', 'log2_norm_score', 'log_norm_score', 'model', 'norm_score', 'observation_schema', 'raw', 'remote_url', 'score', 'score_type', 'summary', 'test', 'url', 'verbose', 'version'])

_Revisiting some code from 1-2 years ago, when (I believe) this used to work fine._

**Comment by appukuttan-shailesh** [comment]
I tested with SciUnit v0.2.3 and the above problem doesn't arise.
Is it a new feature related effect or bug?

**Comment by rgerkin** [comment]
Look into [this method](https://github.com/scidash/sciunit/blob/0c2b13bef8ef2d8222bda8e29041198760820c54/sciunit/base.py#L296) which pickle uses to get the state for pickling.  See if it returns a dict with the contents of related_data intact or not.  If it doesn't, look at the function called inside which gets the list of attributes to hide from the state, based on all the state_hide attributes in all parent classes.  See of that function returns a list with related_data in it or not.

**Comment by appukuttan-shailesh** [comment]
So that method [def __getstate__](https://github.com/scidash/sciunit/blob/0c2b13bef8ef2d8222bda8e29041198760820c54/sciunit/base.py#L296) indeed does not [return](https://github.com/scidash/sciunit/blob/0c2b13bef8ef2d8222bda8e29041198760820c54/sciunit/base.py#L313) `related_data`.

_(I am also curious why we add "state_hide" to `state_hide` [here](https://github.com/scidash/sciunit/blob/0c2b13bef8ef2d8222bda8e29041198760820c54/sciunit/base.py#L313))._

For the example I have shown above, after the above statement, I end up with `state_hide` =
>['temp_dir', 'hash', 'related_data', 'pickling', 'state_hide']

'hash', 'pickling', 'temp_dir' likely arise from [here](https://github.com/scidash/sciunit/blob/7e8635e40d3155a8f8118be6494c67c6c6370899/sciunit/base.py#L293)
and 'state_hide' as mentioned above
'related_data' as you pointed out is added [here](https://github.com/scidash/sciunit/blob/b80d13f52dd1f21c26fe51864c13c411fb644b98/sciunit/scores/base.py#L89)

The problem seems to be that this [line](https://github.com/scidash/sciunit/blob/0c2b13bef8ef2d8222bda8e29041198760820c54/sciunit/base.py#L307) invokes the method '[get_list_attr_with_bases()](https://github.com/scidash/sciunit/blob/0c2b13bef8ef2d8222bda8e29041198760820c54/sciunit/base.py#L421)' which creates the `state_hide` list based on parent class definitions, and does not care for values assigned to the score object by user.

So basically `state_hide` being a class variable (as opposed to an instance variable) currently prevents us from altering it at the score instance level.

**Comment by appukuttan-shailesh** [comment]
@rgerkin Thanks for the merge. would it be possible to release v0.2.7 with this change?

**Comment by rgerkin** [comment]
Done.

**Comment by appukuttan-shailesh** [comment]
Thanks @rgerkin
